Source/WebCore/rendering/RenderBox.cpp

```diff
@@ -1394,6 +1394,75 @@ void RenderBox::clearOverridingLogicalWidthLength()
         gOverridingLogicalWidthLengthMap->remove(this);
 }
 
+std::optional<FlowRelativeDirection> RenderBox::physicalToFlowRelativeDirectionMapping(PhysicalDirection direction, LayoutCalculationPhase layoutCalculationPhase) const
+{
+    // FIXME: Return type should be just FlowRelativeDirection once all code paths have been implemented
+    auto isHorizontalWritingMode = style().isHorizontalWritingMode();
+    auto isLeftToRightDirection = style().isLeftToRightDirection();
+
+    if (layoutCalculationPhase == LayoutCalculationPhase::Positioning) {
+        if (isHorizontalWritingMode == containingBlock()->style().isHorizontalWritingMode()) {
+            switch (direction) {
+            case PhysicalDirection::Top:
+                return isHorizontalWritingMode ? FlowRelativeDirection::BlockStart : (isLeftToRightDirection ? FlowRelativeDirection::InlineStart : FlowRelativeDirection::InlineEnd);
+            default:
+                ASSERT_NOT_IMPLEMENTED_YET();
+            } 
+        }
+    } else
+        ASSERT_NOT_IMPLEMENTED_YET();
+    return { };
+}
+
+MarginTrimType flowRelativeDirectionToMarginTrimType(FlowRelativeDirection direction)
+{
+    switch (direction) {
+    case FlowRelativeDirection::BlockStart:
+        return MarginTrimType::BlockStart;
+    case FlowRelativeDirection::BlockEnd:
+        return MarginTrimType::BlockEnd;
+    case FlowRelativeDirection::InlineStart:
+        return MarginTrimType::InlineStart;
+    case FlowRelativeDirection::InlineEnd:
+        return MarginTrimType::InlineEnd;
+    default:
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+}
+
+void RenderBox::markMarginAsTrimmed(MarginTrimType newTrimmedMargin)
+{
+    ensureRareData().setTrimmedMargins(rareData().trimmedMargins() | static_cast<unsigned>(newTrimmedMargin));
+}
+
+void RenderBox::clearTrimmedMarginsMarkings()
+{
+    if (!hasRareData())
+        return;
+    markMarginAsTrimmed(static_cast<MarginTrimType>(0));
+}
+
+bool RenderBox::hasTrimmedMargin(MarginTrimType marginTrimType) const
+{
+    if (!hasRareData())
+        return false;
+    return rareData().trimmedMargins() & static_cast<unsigned>(marginTrimType);
+}
+
+bool RenderBox::hasTrimmedMargin(PhysicalDirection physicalDirection) const
+{
+    ASSERT(isFlexItem());
+    auto flowRelativeDirection = physicalToFlowRelativeDirectionMapping(physicalDirection, LayoutCalculationPhase::Positioning);
```

Comment:
This is interesting because I'm not actually sure if it's possible to check which phase we are in within render tree is there (i.e. we don't keep track of such a construct during layout)?

For now I added an assert to make sure this is only called on a box that does not need to go through layout since it doesn't really make sure to check for trimmed margins otherwise. I *think* that's a reasonable assertion in this case but let me know if you think otherwise...